project/src/main/java/com/google/autograder/servlets/ListCoursesServlet.java

```diff
@@ -1,66 +1,43 @@
 package com.google.autograder.servlets;
 
-import java.net.URL;
-import java.util.Iterator;
 import java.io.IOException;
-import java.io.BufferedReader;
-import java.io.InputStreamReader;
-import org.json.simple.JSONArray;
 import java.net.HttpURLConnection;
 import org.json.simple.JSONObject;
 import javax.servlet.http.HttpServlet;
-import org.json.simple.parser.JSONParser;
-import java.nio.charset.StandardCharsets;
-import javax.servlet.annotation.WebServlet;
 import com.google.autograder.data.Database;
-import java.io.UnsupportedEncodingException;
-import org.json.simple.parser.ParseException;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import com.google.autograder.data.UserHandler;
-import com.google.appengine.api.datastore.Query;
-import com.google.appengine.api.datastore.Entity;
 import com.google.autograder.servlets.helpers.API;
-import com.google.appengine.api.datastore.Query.Filter;
-import com.google.appengine.api.datastore.PreparedQuery;
-import com.google.appengine.api.datastore.Query.SortDirection;
-import com.google.appengine.api.datastore.Query.FilterOperator;
-import com.google.autograder.servlets.auth.AccessTokenResponse;
-import com.google.appengine.api.datastore.Query.FilterPredicate;
 
 public final class ListCoursesServlet extends HttpServlet {
 
     private static String END_POINT = "https://classroom.googleapis.com/v1/courses";
     
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        if (UserHandler.isUserLoggedIn()) {
-            response.setContentType("application/json");
-
-            String authorization = API.getCurrentUserAPIAuthorization();
-
-            if (authorization != null) {
-                HttpURLConnection connection = (HttpURLConnection) new URL(END_POINT + "?key=" + API.API_KEY).openConnection();
-
-                connection.setRequestMethod("GET");
-                connection.setRequestProperty("courseStates", "ACTIVE");
-                connection.setRequestProperty("Accept", "application/json");
-                connection.setRequestProperty("Authorization", authorization);
+        if (!UserHandler.isUserLoggedIn()) {
+            response.setHeader("redirect", "/index.html");
+            return;
+        }
 
-                String json = API.getJSON(connection);
+        HttpURLConnection connection = API.getAuthenticatedRequest("GET", END_POINT.concat("?key=" + API.API_KEY)).orElse(null);
```

Comment:
Typically when you use Optional you shouldn't need to involve null. The issue Optional fixes when replacing null is forgetting to check for null, which is invalidated if you immediately convert it to null. Instead of this, do something like:
```
Optional<HttpURLConnection> connectionOptional = API.getAuthenticatedRequest(...);
if (!connection.isEmpty()) {
  response.setHeader(...);
  return;
}

HttpURLConnection connection = connectionOptional.get();
```
It's a bit wordier but having a variable declared as Optional makes it immediately clear that it needs to be checked for presence. If you rely on null checking you have to be very careful and remember which methods might return null (like this one) and which ones will not (like getJSON) or check every variable for null

Reply:
+1 Thank you, this makes a lot of sense. Refactoring now to more properly utilize the Optional type and avoid null checking.